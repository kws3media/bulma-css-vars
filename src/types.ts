export interface BulmaCssVarsOptions {
  sassOutputFile: string
  sassEntryFile: string
  colorDefs: { [colorName: string]: ColorDef }
}

export interface Hsl {
  h: number
  s: number
  l: number
  a?: number
}
export interface Rgb {
  r: number
  g: number
  b: number
  a?: number
}
export type Hex = string

export type ColorDef = Hsl | Rgb | Hex

export type ColorFn =
  | 'rgba'
  | 'adjusthue'
  | 'saturate'
  | 'desaturate'
  | 'lighten'
  | 'darken'
  | 'color-invert'
  | 'dark-color'
  | 'light-color'

export interface ColorFnCall {
  fn: ColorFn
  fnArg: string
  composeArg?: ColorFnCall
}

export interface ColorCallSet {
  [color: string]: {
    calls: ColorFnCall[]
    value?: Hsl
  }
}

export interface NameValueColor {
  name: string
  value: string
}
