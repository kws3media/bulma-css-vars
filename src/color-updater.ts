import {
  bulmaColorTools,
  getNameValFromColorDef,
  stringToHsl,
  strValFromColorDef,
} from './bulma-color-tools'
import { ColorFn, ColorFnCall, ColorCallSet, NameValueColor } from './types'

export class ColorUpdater {
  constructor(protected colorVals: ColorCallSet) {}

  protected getBaseVarNames() {
    return Object.keys(this.colorVals)
  }
  protected getBaseVars() {
    return Object.entries(this.colorVals).map(([name, { value: hsl }]) =>
      getNameValFromColorDef(name, hsl)
    )
  }
  protected getDerivedVars() {
    return Object.entries(this.colorVals)
      .filter(([_base, { calls }]) => calls.length)
      .map(([base, { calls }]) => {
        return calls.map((call) => this.callToNameVal(base, call))
      })
  }
  getUpdatedVars(colorName: string, colorVal: string) {
    if (!(colorName in this.colorVals)) {
      console.warn(`Color '${colorName}' was not configured in bulma-css-vars!`)
      return []
    }
    const value = stringToHsl(colorVal)
    this.colorVals[colorName].value = value
    const calls =
      (this.colorVals[colorName] && this.colorVals[colorName].calls) || []
    return [
      getNameValFromColorDef(colorName, value),
      ...calls.map((call) => this.callToNameVal(colorName, call)),
    ]
  }

  updateVarsInDocument(colorName: string, colorVal: string) {
    const updateds = this.getUpdatedVars(colorName, colorVal)
    updateds.forEach(({ name, value }) =>
      document.documentElement.style.setProperty(name, value)
    )
  }

  protected callToNameVal(base: string, call: ColorFnCall) {
    const name = `--${base}${this.createDerivedVarName(call)}`
    const value = this.createDerivedVariable(base, call)
    return { name, value }
  }

  protected createDerivedVariable(baseColorName: string, call: ColorFnCall) {
    if (!(baseColorName in this.colorVals)) {
      throw Error(
        `Color '${baseColorName}' is not defined! Cannot derive required colors!`
      )
    }
    function getEvaluatedColorFromFnCall(
      fnName: ColorFn,
      fnArg: string,
      colorArg: string,
      compose?: ColorFnCall
    ): string {
      const evaluated = bulmaColorTools[fnName](colorArg, fnArg)
      if (!compose) {
        return evaluated
      }
      return getEvaluatedColorFromFnCall(
        compose.fn,
        compose.fnArg,
        evaluated,
        compose.composeArg
      )
    }

    const colorVal = this.colorVals[baseColorName].value
    const baseColor = strValFromColorDef(colorVal, baseColorName)
    return getEvaluatedColorFromFnCall(
      call.fn,
      call.fnArg,
      baseColor,
      call.composeArg
    )
  }

  protected createDerivedVarName(call: ColorFnCall) {
    function evaluateCallName(
      fnName: ColorFn,
      fnArg: string,
      compose?: ColorFnCall
    ): string {
      const partCompose = compose
        ? evaluateCallName(compose.fn, compose.fnArg, compose.composeArg)
        : ''
      const partFnArg = fnArg ? `--${fnArg}` : ''
      return `${partFnArg}--${fnName}${partCompose}`
    }
    return evaluateCallName(call.fn, call.fnArg, call.composeArg)
  }
}

export class ColorGenerator extends ColorUpdater {
  getAllVars() {
    return [
      ...this.getBaseVars(),
      ...([] as NameValueColor[]).concat(...this.getDerivedVars()),
    ]
  }

  createWritableSassFileOnlySassBaseVariables(): string {
    return `${this.getBaseVarNames()
      .map((name) => `$${name}: var(--${name})`)
      .join('\n')}
`
  }

  createWritableSassFile(): string {
    const baseSassVariableStyles =
      this.createWritableSassFileOnlySassBaseVariables()

    const baseCssVariableStyles = `#{":root"}
${this.getBaseVars()
  .map(({ name, value }) => `  ${name}: ${value}`)
  .join('\n')}
`

    const derivedCssVarStyles = `#{":root"}
${this.getDerivedVars()
  .map((vars) =>
    vars.map(({ name, value }) => `  ${name}: ${value}`).join('\n')
  )
  .join('\n\n')}
`

    const fullFile = `
// sass variables
${baseSassVariableStyles}

// declared base css variables
${baseCssVariableStyles}

// derived, generated css variables
${derivedCssVarStyles}
`

    return fullFile
  }
}
