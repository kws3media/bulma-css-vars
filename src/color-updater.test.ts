import { ColorUpdater, ColorGenerator } from './color-updater'
import { ColorCallSet } from './types'

const colorCallSet: ColorCallSet = {
  black: {
    value: { h: 42, s: 42, l: 42 },
    calls: [
      {
        fn: 'adjusthue',
        fnArg: '42deg',
        composeArg: {
          fn: 'lighten',
          fnArg: '4200', // 42%
        },
      },
      {
        fn: 'color-invert',
        fnArg: null,
      },
    ],
  },
}

describe('Color Updater', () => {
  test('adds updates for all color defs', () => {
    const updater = new ColorUpdater(colorCallSet)
    const updates = updater.getUpdatedVars('black', '#5291a3')
    expect(updates).toEqual([
      { name: '--black', value: 'rgb(82, 145, 163)' },
      {
        name: '--black--42deg--adjusthue--4200--lighten',
        value: 'rgb(221, 222, 238)',
      },
      { name: '--black--color-invert', value: 'rgb(255, 255, 255)' },
    ])
  })

  test('lets updater set updates on document element style', () => {
    // mock global here for a fake browser env - it's only required for this test
    const g: any = global
    g['document'] = {
      documentElement: {
        style: {
          setProperty(_name: string, _value: string) {},
        },
      },
    }

    const spySetStyle = jest.spyOn(
      document.documentElement.style,
      'setProperty'
    )
    const updater = new ColorUpdater(colorCallSet)
    updater.updateVarsInDocument('black', '#5291a3')
    expect(spySetStyle).toHaveBeenCalled()
    expect(spySetStyle).toHaveBeenCalledTimes(3)
    expect(spySetStyle.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "--black",
          "rgb(82, 145, 163)",
        ],
        Array [
          "--black--42deg--adjusthue--4200--lighten",
          "rgb(221, 222, 238)",
        ],
        Array [
          "--black--color-invert",
          "rgb(255, 255, 255)",
        ],
      ]
    `)
  })
})

describe('Color Generator', () => {
  test('writes base variables in sass style', () => {
    const generator = new ColorGenerator(colorCallSet)
    const sassBaseVars = generator.createWritableSassFileOnlySassBaseVariables()
    expect(sassBaseVars).toMatchInlineSnapshot(`
      "$black: var(--black)
      "
    `)
  })

  test('writes all variables in sass style', () => {
    const generator = new ColorGenerator(colorCallSet)
    const sassVars = generator.createWritableSassFile()
    expect(sassVars).toMatchInlineSnapshot(`
      "
      // sass variables
      $black: var(--black)


      // declared base css variables
      #{\\":root\\"}
        --black: rgb(82, 145, 163)


      // derived, generated css variables
      #{\\":root\\"}
        --black--42deg--adjusthue--4200--lighten: rgb(221, 222, 238)
        --black--color-invert: rgb(255, 255, 255)

      "
    `)
  })
})
