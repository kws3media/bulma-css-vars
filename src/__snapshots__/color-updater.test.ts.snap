// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color Generator adds custom block wrapper: "[data-theme="dark"] {
  --black: rgb(82, 145, 163);
  --black--42deg--adjusthue--4200--lighten: rgb(221, 222, 238);
  --black--color-invert: rgb(255, 255, 255);
}" 1`] = `
"[data-theme=\\"dark\\"] {
  --black: rgb(82, 145, 163);
  --black--42deg--adjusthue--4200--lighten: rgb(221, 222, 238);
  --black--color-invert: rgb(255, 255, 255);
}"
`;

exports[`Color Generator writes all variables in scss style: ":root {
  --black: rgb(82, 145, 163);
  --black--42deg--adjusthue--4200--lighten: rgb(221, 222, 238);
  --black--color-invert: rgb(255, 255, 255);
}" 1`] = `
":root {
  --black: rgb(82, 145, 163);
  --black--42deg--adjusthue--4200--lighten: rgb(221, 222, 238);
  --black--color-invert: rgb(255, 255, 255);
}"
`;
